### DISCUSSION TITLE
`z.discriminatedUnion` with object as `z.literal`

### DISCUSSION THREAD

**Discussion by pablopsp** [colinhacks/zod#3213] [Q&A] [answered]
I'm developing a form using the `react-hook-form` library. In my form, I've incorporated a `react-select` component to handle specific options. This component acts as a discriminator for validation, meaning that depending on the selected option, the form validation varies.
I've been exploring `zod` and encountered a challenge related to using an object as a `z.literal` for a `z.discriminatedUnion`.

Despite my efforts to resolve it by referring to the documentation and previous discussions, I haven't been able to find a satisfactory solution.

Here is an example snippet of code that illustrates a similar structure to what I've been working on:
```js
import z from "zod";

const options = [
  { label: "Option A", value: "optionA" },
  { label: "Option B", value: "optionB" },
];

const optionASchema = z.object({
  category: z.literal(options[0]),
  attributeA: z.boolean(),
});

const optionBSchema = z.object({
  category: z.literal(options[1]),
  attributeB: z.boolean(),
});

const categoryDiscriminated = z.discriminatedUnion("category", [
  optionASchema,
  optionBSchema,
]);

const categoryIntersection = z.intersection(
  z.object({
    category: z.object(
      {
        label: z.string(),
        value: z.any(),
      },
      {
        required_error: "Category is required",
      }
    ),
  }),
  categoryDiscriminated
);

const schema = z.object({
  firstName: z.string(),
  lastName: z.string(),
  details: categoryIntersection,
});

const parseA = schema.safeParse({
  firstName: "John",
  lastName: "Doe",
  details: {
    category: options[0],
    attributeA: false,
  },
});
console.log(parseA);

const parseB = schema.safeParse({
  firstName: "John",
  lastName: "Doe",
  details: {
    category: { label: "Option A", value: "optionA" },
    attributeA: false,
  },
});
console.log(parseB);
```

I believe the issue stems from comparing objects by reference. 
I would prefer to utilize the value property of the `react-select` option only if using the entire object proves to be not viable. Any assistance or guidance on this matter would be greatly appreciated.

Thank you for your time and support.

**Comment by JacobWeisenburger** [comment]
> I believe the issue stems from comparing objects by reference

I'm pretty sure that's correct, you need the discriminator to be a value and not a reference.

**Comment by pablopsp** [comment]
I've been trying to use `z.preprocess` along with `.pipe()` to extract the value for `z.literal`, but so far, I haven't been successful. 
Is it possible to validate the selected option of `react-select` and use its value as the key for `z.literal` in `z.discriminatedUnion`?

**Reply by JacobWeisenburger** [reply]
I'm unfamiliar with how `react-select` works, so if you provide a full example maybe I can take a look

**Reply by pablopsp** [reply]
Here's an example of it https://stackblitz.com/edit/stackblitz-starters-y9bema

**Reply by JacobWeisenburger** [reply]
this has nothing to do with zod. you need to do something like this:
```tsx
<Controller
  name="details.category"
  control={control}
  render={({ field }) => (
    <Select
      {...field}
      options={options}
      isClearable
      onChange={(e) => {
        form.setValue('details.category', e.value);
      }}
    />
  )}
/>
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by pablopsp** [reply]
The answer is not viable since the `react-select` value must conform to the `{ label: any, value: any }` structure if I just use the value the select wouldn't change. I am investigating whether Zod offers a solution that enables a direct modification of the value

```js
const optionASchema = z.object({
  category: z.literal(options[0]), // Use options[0].value as z.literal but receiving { label: any, value: any }
  attributeA: z.boolean(),
});
```

**Reply by JacobWeisenburger** [reply, marked as answer]
I don't think that is possible.

**Reply by pablopsp** [reply]
Think so, anyway, thank you for your assistance and the quick answers 🙂
